vllm/entrypoints/openai/controller.py

```diff
@@ -0,0 +1,351 @@
+        return True
+
+    def speed_normalize(self, queue_length: int):
+        """For the worker, the more it has unfisinished queue,
+        it means that it has less capacity for serving future request.
+
+        Thus we here use 1/(queue_length+1) for load balance algo.
+        """
+        return 1. / (queue_length + 1)
+
+    def remove_worker(self, worker_addr: str):
+        logger.info("Remove stale worker: %s", worker_addr)
+        del self.worker_info[worker_addr]
+
+    def list_models(self):
+        model_names = []
+
+        for w_name, w_info in self.worker_info.items():
+            for model in w_info.model_names:
+                model_names.append(model)
+
+        return sorted(set(model_names))
+
+    def list_workers(self):
+        cur = time.time()
+        workers = {}
+
+        for w_name, w_info in self.worker_info.items():
+            worker = {
+                w_name: {
+                    "model": w_info.model_names,
+                    "req_cnt": w_info.req_cnt,
+                    "speed": w_info.speed,
+                    "check_heart_beat": w_info.check_heart_beat,
+                    "last_heart_beat": cur - w_info.last_heart_beat,
+                }
+            }
+            workers.update(worker)
+
+        return workers
+
+    def get_worker_address(self, model_name: str):
+        worker_names = []
+        worker_qlen = []
+        for w_name, w_info in self.worker_info.items():
+            worker_names.append(w_name)
+            worker_qlen.append(w_info.req_cnt / w_info.speed)
+        if len(worker_names) == 0:
+            return ""
+
+        # return the worker with least normalized queue length
+        min_index = np.argmin(worker_qlen)
+        w_name = worker_names[min_index]
+        self.worker_info[w_name].req_cnt += 1
+
+        return w_name
+
+    def receive_heart_beat(self, worker_addr: str, queue_length: int):
+        if worker_addr not in self.worker_info:
+            logger.info("Receive unknown heart beat. %s", worker_addr)
+            return False
+
+        speed = self.speed_normalize(queue_length)
+        self.worker_info[worker_addr].speed = speed
+        self.worker_info[worker_addr].last_heart_beat = time.time()
+        logger.info("Receive heart beat. %s speed %f", worker_addr, speed)
+        return True
+
+    def remove_stale_workers_by_expiration(self):
+        expire = time.time() - CONTROLLER_HEART_BEAT_EXPIRATION
+        to_delete = []
+        for worker_addr, w_info in self.worker_info.items():
+            if w_info.check_heart_beat and w_info.last_heart_beat < expire:
+                to_delete.append(worker_addr)
+
+        for worker_addr in to_delete:
+            self.remove_worker(worker_addr)
+
+
+def heart_beat_worker(obj):
+    while True:
+        time.sleep(WORKER_HEART_BEAT_INTERVAL)
+        obj.send_heart_beat()
+
+
+class Worker:
+
+    def __init__(self, controller_addr, worker_ip, worker_port, model_names,
```

Comment:
Can you fill in the type annotations for the initializer function?

Reply:
Thanks for point it out, I fix it in latest commit